### ISSUE TITLE
Help linking vdj to seurat object

### ISSUE THREAD

**Issue by JFavaloro** [ncborcherding/scRepertoire#22]
Hi,

Thanks so much for this tool - it is exactly what I was after for my analysis!

I'm hoping to get some help with linking clonotype data to seurat objects (I am very new to R - apologies!).

I have paired gene expression and vdj data from 3 patients / 2 tissues. I have created 6 merged seurat objects and associated clonotypes as illustrated below:

    

Patient | TIssue | Object 1 | Object 2 | Object 3 | Object 4 | Object 5 | Object 6
-- | -- | -- | -- | -- | -- | -- | --
1 | A | X |   |   | X |   | X
  | B | X |   |   |   | X | X
2 | A |   | X |   | X |   | X
  | B |   | X |   |   | X | X
3 | A |   |   | X | X |   | X
  | B |   |   | X |   | X | X

I'm using the code (for example):

object6.combined = combineExpression(object6_vdj, object6.combined, cloneCall = "gene", groupBy = "sample")

Which works great for Objects 4, 5 and 6, however I'm getting the error:

Error in `$<-.data.frame`(`*tmp*`, "cloneType", value = NA):
 replacement has 1 row, data has 0

When trying link clonotype and seurat data for objects 1, 2 and 3.

I'm assuming this is due to the groupBy argument as 'sample' is one value for these objects (e.g. for Object 1, sample = Patient 1), however changing to another value (e.g. ID, which in this case is Tissue (A or B)) gives the same error. Removing the groupBy argument completely transfers NA values across the board (barcode, CTgene, CTnt, CTaa, CTstrict, Frequency, cloneType).

Any help would be greatly appreciated and again, thank you so much for this package!

**Comment by ncborcherding** [comment]
Hey JFavaloro,

Thanks for the post. 

My first question would be is there a varying name strategy for your Seurat objects? I get a very similar error if my vdj object does not match my seurat object cell names. If you could give the output of **head(objectX.combined[[]]))** that would be help rule that out. 

What is the code you are using to combine the vdjs? Does that vary between your combinations? 

Removing the groupBy argument from combineExpression() will lead to the frequency calculation being based on the individual list element of the combined VDJ - so in your data Object 1 - there will be separate frequencies based on presence in the Tissue A versus Tissue B.

Hope that helps, feel free to send the info and we can figure this out.

Thanks,
Nick

**Comment by JFavaloro** [comment]
Hi Nick,

Thanks so much for the quick response!

I don't think there is a varying name strategy - as I mentioned, combineExpression works when there is more than one patient sample in a combined object.

Here's the object head from Object 6 (one that works) trimmed  to the top row so the post doesn't get too clustered:

  | orig.ident | nCount_RNA | nFeature_RNA | percent.mt | integrated_snn_res.0.5 | seurat_clusters | barcode | CTgene | CTnt | CTaa | CTstrict | Frequency | cloneType
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
a1_AGTTGGTAGCCTCGTG-1 | a1 | 1151 | 603 | 3.99652476 | 7 | 7 | a1_AGTTGGTAGCCTCGTG-1 | TRAV29/DV5.TRAJ45.TRAC_TRBV28.TRBJ1-2.None.TRBC1 | TGTGCAGCAAGCGGAGTGTATTCAGGAGGAGGTGCTGACGGACTCACCTTT_TGTGCCACCGGACCGGGACAGCCTATATTGGGCTACACCTTC | CAASGVYSGGGADGLTF_CATGPGQPILGYTF | TRAV29/DV5.TRAJ45.TRAC_TGTGCAGCAAGCGGAGTGTATTCAGGAGGAGGTGCTGACGGACTCACCTTT_TRBV28.TRBJ1-2.None.TRBC1_TGTGCCACCGGACCGGGACAGCCTATATTGGGCTACACCTTC | 2 | Small (1 < X <= 5)

My workflow is:
1. Load 10x GEX data
2. Load filtered_contig_annotations with stringAsFactors = FALSE
3. Combine filtered filtered_contig_annotations into the 6 objects as above using the code (for example - Object 1 (identical to objects 2 and 3 just with different patients)

pt1_vdj = list(a1_vdj, b1_vdj)
pt1_vdj <- combineTCR(pt1_vdj, 
                      samples = c("pt1", "pt1"), 
                      ID = c("a", "b"), cells ="T-AB")
pt1_vdj[["pt1_a"]][["barcode"]] <- gsub('pt1_a', 'a1', pt1_vdj[["pt1_a"]][["barcode"]])
pt1_vdj[["pt1_b"]][["barcode"]] <- gsub('pt1_b', 'b1', pt1_vdj[["pt1_b"]][["barcode"]])

NB: I change the barcodes to match the Seurat objects naming convention. The code for objects 4-6 is identical except I'm merging files from more than one patient - so for example, Object 4:

a_vdj = list(a1_vdj, a2_vdj, a3_vdj)
a_vdj <- combineTCR(a_vdj, 
                     samples = c("pt1", "pt2", "pt3"), 
                     ID = c("a", "a", "a"), cells ="T-AB")
a_vdj[["pt1_a"]][["barcode"]] <- gsub('pt1_a', 'a1', a_vdj[["pt1_a"]][["barcode"]])
a_vdj[["pt2_a"]][["barcode"]] <- gsub('pt2_a', 'a2', a_vdj[["pt2_a"]][["barcode"]])
a_vdj[["pt3_a"]][["barcode"]] <- gsub('pt3_a', 'a3', a_vdj[["pt3_a"]][["barcode"]])

4. Create the Seurat objects
5. Merge Seurat objects into the 6 objects as above
6. Cluster - creating the x.combined object
7. Combine clustered seurat objects with the corresponding vdj from scRepertoire using the code (for example - Object 1)

pt1.combined = combineExpression(pt1_vdj, pt1.combined, cloneCall="gene", groupBy = "sample", cloneTypes = c(None = 0, Single = 1, Small = 5, Medium = 20, Large = 200, Hyperexpanded = 2000))

Which gives the error I've mentioned above. 

The code for objects 4-6 is identical (just replace pt1 with 'a' etc) and works great!

Thanks again!

**Comment by ncborcherding** [comment]
Thank you for the thorough run down, very much appreciated. I see and like how you are modifying the VDJ barcodes - very smart. 

What are the barcode patterns for you Objects 1-3? A way to tell the if your barcodes match might be looking at the intersection of the cell names in your Seurat object and the barcodes for the VDJ list elements? Depending on the set up - there Seurat might add a different hyphenated suffix value as well. There is a filtering step when grouping by sample that would produce any empty tmp object (Con.df) that when trying to add the variable **cloneType**, would produce the error you are seeing. 

Here is the specific part of the code I am referring to (starts on line 61 in the seuratFunctions.R).
```
else if (groupBy != "none") {
        data <- data.frame(bind_rows(df), stringsAsFactors = FALSE)
        data2 <- na.omit(unique(data[,c("barcode", cloneCall, groupBy)]))
        data2 <- data2[data2[,"barcode"] %in% cell.names, ] #here cell.names denotes the exact barcode of the Seurat object
        data2 <- as.data.frame(data2 %>% group_by(data2[,cloneCall], 
                    data2[,groupBy]) %>% summarise(Frequency = n()))
        colnames(data2)[c(1,2)] <- c(cloneCall, groupBy)
        x <- unique(data[,groupBy])
        for (i in seq_along(x)) {
            sub1 <- subset(data, data[,groupBy] == x[i])
            sub2 <- subset(data2, data2[,groupBy] == x[i])
            merge <- merge(sub1, sub2, by=cloneCall)
            Con.df <- rbind.data.frame(Con.df, merge) } }
    Con.df$cloneType <- NA #Here is where the error is occurring
```
If this is not the case, let me know and we can troubleshoot more. Sorry for the inconvenience.

**Comment by JFavaloro** [comment]
Hi Nick,

Thanks again for such a quick response and the kind words! No inconvenience at all!

Happy to report I've solved it! It was carelessness on my behalf and you are right - it was an issue with matching the barcodes.

I went back through my code and found the error; when I created the merged objects I neglected to at the patient ID to add.cell.ids - so the barcodes for object 1-3 only consisted of tissue

For anyone else who has a similar issue I changed:

pt1 <- merge(a1, y = b1, add.cell.ids = c("a", "b"), project = "pt1")

to

pt1 <- merge(a1, y = b1, add.cell.ids = c("a**1**", "b**1**"), project = "pt1")

Thanks so much for the help and again - thanks so much for scRepertoire - fantastic package 👍
